python/pyiceberg/avro/reader.py

```diff
@@ -238,41 +249,50 @@ def skip(self, decoder: BinaryDecoder) -> None:
             return self.option.skip(decoder)
 
 
-class StructProtocolReader(Reader):
-    create_struct: Callable[[], StructProtocol]
-    fields: Tuple[Tuple[Optional[int], Reader], ...]
+class StructReader(Reader):
+    field_readers: Tuple[Tuple[Optional[int], Reader], ...]
+    create_struct: Type[StructProtocol]
+    struct: Optional[StructType]
 
-    def __init__(self, fields: Tuple[Tuple[Optional[int], Reader], ...], create_struct: Callable[[], StructProtocol]):
-        self.create_struct = create_struct
-        self.fields = fields
-
-    def create_or_reuse(self, reuse: Optional[StructProtocol]) -> StructProtocol:
-        if reuse:
-            return reuse
-        else:
-            return self.create_struct()
+    def __init__(
+        self,
+        field_readers: Tuple[Tuple[Optional[int], Reader], ...],
+        create_struct: Optional[Type[StructProtocol]] = None,
```

Comment:
This creates an awkward situation where the struct is always passed to the type's constructor. If we used `Callable[[StructType], StructProtocol]` instead then the caller could pass in a lambda that ignores the struct type, rather than adding requirements on the user's type.

Reply:
Of course. Keep in mind that this is orthogonal to not wanting to define the constructor in the `StructProtocol`, this requires the user to pass in `__init__(struct: StructType)`. Therefore I had to refactor the tests, since `Record(a=1, b="abc")` also isn't allowed.

Reply:
Inspecting gives different results amongst different Python version 😭  It works fine locally on 3.10, but CI fails on 3.8.

Looks like it doesn't look up the constructor of the class that it inherits from:
```
<class 'pyiceberg.avro.file.AvroFileHeader'>
FullArgSpec(args=['cls'], varargs='args', varkw='kwds', defaults=None, kwonlyargs=[], kwonlydefaults=None, annotations={})
```

Reply:
Moving to an _it’s easier to ask for forgiveness than permission_ approach